build failure on upcoming gcc-12: test/src/unit-regression1.cpp:392:22: error: ambiguous overload for 'operator='

On `gcc-12` (20211107 snapshot) `std::string` amended it's constructors to a form that fails regression tests building:

```
mkdir b && cd b
cmake ..
make
...
[ 71%] Building CXX object test/CMakeFiles/test-regression1.dir/src/unit-regression1.cpp.o
/home/slyfox/dev/git/nlohmann-json/test/src/unit-regression1.cpp: In function 'void _DOCTEST_ANON_FUNC_2()':
/home/slyfox/dev/git/nlohmann-json/test/src/unit-regression1.cpp:392:22: error: ambiguous overload for 'operator=' (operand types are 'std::string' {aka 'std::__cxx11::basic_string<char>'} and 'nlohmann::basic_json<>::value_type' {aka 'nlohmann::basic_json<>'})
  392 |         s2 = o["name"];
      |                      ^
In file included from /nix/store/0hvbmcf38gqpr6c49h7zl3iy228afaz0-gcc-12.0.0/include/c++/12.0.0/string:53,
                 from /nix/store/0hvbmcf38gqpr6c49h7zl3iy228afaz0-gcc-12.0.0/include/c++/12.0.0/bits/locale_classes.h:40,
                 from /nix/store/0hvbmcf38gqpr6c49h7zl3iy228afaz0-gcc-12.0.0/include/c++/12.0.0/locale:39,
                 from /home/slyfox/dev/git/nlohmann-json/test/src/unit-regression1.cpp:33:
/nix/store/0hvbmcf38gqpr6c49h7zl3iy228afaz0-gcc-12.0.0/include/c++/12.0.0/bits/basic_string.h:627:21: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(std::nullptr_t) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>; std::nullptr_t = std::nullptr_t]' (deleted)
  627 |       basic_string& operator=(nullptr_t) = delete;
      |                     ^~~~~~~~
/nix/store/0hvbmcf38gqpr6c49h7zl3iy228afaz0-gcc-12.0.0/include/c++/12.0.0/bits/basic_string.h:682:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(const std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
  682 |       operator=(const basic_string& __str)
      |       ^~~~~~~~
/nix/store/0hvbmcf38gqpr6c49h7zl3iy228afaz0-gcc-12.0.0/include/c++/12.0.0/bits/basic_string.h:720:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
  720 |       operator=(basic_string&& __str)
      |       ^~~~~~~~
In file included from /home/slyfox/dev/git/nlohmann-json/test/thirdparty/doctest/doctest_compatibility.h:6,
                 from /home/slyfox/dev/git/nlohmann-json/test/src/unit-regression1.cpp:30:
/home/slyfox/dev/git/nlohmann-json/test/src/unit-regression1.cpp:398:37: error: ambiguous overload for 'operator=' (operand types are 'std::string' {aka 'std::__cxx11::basic_string<char>'} and 'nlohmann::basic_json<>::value_type' {aka 'nlohmann::basic_json<>'})
  398 |         CHECK_THROWS_AS(s2 = o["int"], json::type_error);
      |                                     ^
/home/slyfox/dev/git/nlohmann-json/test/thirdparty/doctest/doctest.h:1844:23: note: in definition of macro 'DOCTEST_CAST_TO_VOID'
 1844 |     static_cast<void>(__VA_ARGS__);                                                                \
      |                       ^~~~~~~~~~~
/home/slyfox/dev/git/nlohmann-json/test/thirdparty/doctest/doctest.h:2163:44: note: in expansion of macro 'DOCTEST_ASSERT_THROWS_AS'
 2163 | #define DOCTEST_CHECK_THROWS_AS(expr, ...) DOCTEST_ASSERT_THROWS_AS(expr, DT_CHECK_THROWS_AS, "", __VA_ARGS__)
      |                                            ^~~~~~~~~~~~~~~~~~~~~~~~
/home/slyfox/dev/git/nlohmann-json/test/thirdparty/doctest/doctest.h:2576:36: note: in expansion of macro 'DOCTEST_CHECK_THROWS_AS'
 2576 | #define CHECK_THROWS_AS(expr, ...) DOCTEST_CHECK_THROWS_AS(expr, __VA_ARGS__)
      |                                    ^~~~~~~~~~~~~~~~~~~~~~~
/home/slyfox/dev/git/nlohmann-json/test/src/unit-regression1.cpp:398:9: note: in expansion of macro 'CHECK_THROWS_AS'
  398 |         CHECK_THROWS_AS(s2 = o["int"], json::type_error);
      |         ^~~~~~~~~~~~~~~
In file included from /nix/store/0hvbmcf38gqpr6c49h7zl3iy228afaz0-gcc-12.0.0/include/c++/12.0.0/string:53,
                 from /nix/store/0hvbmcf38gqpr6c49h7zl3iy228afaz0-gcc-12.0.0/include/c++/12.0.0/bits/locale_classes.h:40,
                 from /nix/store/0hvbmcf38gqpr6c49h7zl3iy228afaz0-gcc-12.0.0/include/c++/12.0.0/locale:39,
                 from /home/slyfox/dev/git/nlohmann-json/test/src/unit-regression1.cpp:33:
/nix/store/0hvbmcf38gqpr6c49h7zl3iy228afaz0-gcc-12.0.0/include/c++/12.0.0/bits/basic_string.h:627:21: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(std::nullptr_t) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>; std::nullptr_t = std::nullptr_t]' (deleted)
  627 |       basic_string& operator=(nullptr_t) = delete;
      |                     ^~~~~~~~
/nix/store/0hvbmcf38gqpr6c49h7zl3iy228afaz0-gcc-12.0.0/include/c++/12.0.0/bits/basic_string.h:682:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(const std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
  682 |       operator=(const basic_string& __str)
      |       ^~~~~~~~
/nix/store/0hvbmcf38gqpr6c49h7zl3iy228afaz0-gcc-12.0.0/include/c++/12.0.0/bits/basic_string.h:720:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
  720 |       operator=(basic_string&& __str)
      |       ^~~~~~~~
In file included from /home/slyfox/dev/git/nlohmann-json/test/thirdparty/doctest/doctest_compatibility.h:6,
                 from /home/slyfox/dev/git/nlohmann-json/test/src/unit-regression1.cpp:30:
/home/slyfox/dev/git/nlohmann-json/test/src/unit-regression1.cpp:402:39: error: ambiguous overload for 'operator=' (operand types are 'std::string' {aka 'std::__cxx11::basic_string<char>'} and 'nlohmann::basic_json<>::value_type' {aka 'nlohmann::basic_json<>'})
  402 |         CHECK_THROWS_WITH(s2 = o["int"], "[json.exception.type_error.302] type must be string, but is number");
      |                                       ^
/home/slyfox/dev/git/nlohmann-json/test/thirdparty/doctest/doctest.h:1844:23: note: in definition of macro 'DOCTEST_CAST_TO_VOID'
 1844 |     static_cast<void>(__VA_ARGS__);                                                                \
      |                       ^~~~~~~~~~~
/home/slyfox/dev/git/nlohmann-json/test/thirdparty/doctest/doctest.h:2167:46: note: in expansion of macro 'DOCTEST_ASSERT_THROWS_WITH'
 2167 | #define DOCTEST_CHECK_THROWS_WITH(expr, ...) DOCTEST_ASSERT_THROWS_WITH(expr, #expr, DT_CHECK_THROWS_WITH, __VA_ARGS__)
      |                                              ^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/slyfox/dev/git/nlohmann-json/test/thirdparty/doctest/doctest.h:2577:38: note: in expansion of macro 'DOCTEST_CHECK_THROWS_WITH'
 2577 | #define CHECK_THROWS_WITH(expr, ...) DOCTEST_CHECK_THROWS_WITH(expr, __VA_ARGS__)
      |                                      ^~~~~~~~~~~~~~~~~~~~~~~~~
/home/slyfox/dev/git/nlohmann-json/test/src/unit-regression1.cpp:402:9: note: in expansion of macro 'CHECK_THROWS_WITH'
  402 |         CHECK_THROWS_WITH(s2 = o["int"], "[json.exception.type_error.302] type must be string, but is number");
      |         ^~~~~~~~~~~~~~~~~
In file included from /nix/store/0hvbmcf38gqpr6c49h7zl3iy228afaz0-gcc-12.0.0/include/c++/12.0.0/string:53,
                 from /nix/store/0hvbmcf38gqpr6c49h7zl3iy228afaz0-gcc-12.0.0/include/c++/12.0.0/bits/locale_classes.h:40,
                 from /nix/store/0hvbmcf38gqpr6c49h7zl3iy228afaz0-gcc-12.0.0/include/c++/12.0.0/locale:39,
                 from /home/slyfox/dev/git/nlohmann-json/test/src/unit-regression1.cpp:33:
/nix/store/0hvbmcf38gqpr6c49h7zl3iy228afaz0-gcc-12.0.0/include/c++/12.0.0/bits/basic_string.h:627:21: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(std::nullptr_t) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>; std::nullptr_t = std::nullptr_t]' (deleted)
  627 |       basic_string& operator=(nullptr_t) = delete;
      |                     ^~~~~~~~
/nix/store/0hvbmcf38gqpr6c49h7zl3iy228afaz0-gcc-12.0.0/include/c++/12.0.0/bits/basic_string.h:682:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(const std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
  682 |       operator=(const basic_string& __str)
      |       ^~~~~~~~
/nix/store/0hvbmcf38gqpr6c49h7zl3iy228afaz0-gcc-12.0.0/include/c++/12.0.0/bits/basic_string.h:720:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
  720 |       operator=(basic_string&& __str)
      |       ^~~~~~~~
/home/slyfox/dev/git/nlohmann-json/test/src/unit-regression1.cpp:1174:16: error: ambiguous overload for 'operator=' (operand types are 'std::string' {aka 'std::__cxx11::basic_string<char>'} and 'nlohmann::json' {aka 'nlohmann::basic_json<>'})
 1174 |         test = v;
      |                ^
In file included from /nix/store/0hvbmcf38gqpr6c49h7zl3iy228afaz0-gcc-12.0.0/include/c++/12.0.0/string:53,
                 from /nix/store/0hvbmcf38gqpr6c49h7zl3iy228afaz0-gcc-12.0.0/include/c++/12.0.0/bits/locale_classes.h:40,
                 from /nix/store/0hvbmcf38gqpr6c49h7zl3iy228afaz0-gcc-12.0.0/include/c++/12.0.0/locale:39,
                 from /home/slyfox/dev/git/nlohmann-json/test/src/unit-regression1.cpp:33:
/nix/store/0hvbmcf38gqpr6c49h7zl3iy228afaz0-gcc-12.0.0/include/c++/12.0.0/bits/basic_string.h:627:21: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(std::nullptr_t) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>; std::nullptr_t = std::nullptr_t]' (deleted)
  627 |       basic_string& operator=(nullptr_t) = delete;
      |                     ^~~~~~~~
/nix/store/0hvbmcf38gqpr6c49h7zl3iy228afaz0-gcc-12.0.0/include/c++/12.0.0/bits/basic_string.h:682:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(const std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
  682 |       operator=(const basic_string& __str)
      |       ^~~~~~~~
/nix/store/0hvbmcf38gqpr6c49h7zl3iy228afaz0-gcc-12.0.0/include/c++/12.0.0/bits/basic_string.h:720:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
  720 |       operator=(basic_string&& __str)
      |       ^~~~~~~~
make[2]: *** [test/CMakeFiles/test-regression1.dir/build.make:76: test/CMakeFiles/test-regression1.dir/src/unit-regression1.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:2148: test/CMakeFiles/test-regression1.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
```

#### Which compiler and operating system are you using?

- Compiler: gcc-12 development snapshot
- Operating system: x86_64-linux

#### Which version of the library did you use?

- [x] the `develop` branch

Thanks @t-b for the fix! I tried it, and it worked for GCC 12. Unfortunately, it does not work for any other compiler :-/

See https://github.com/nlohmann/json/pull/3226.

The error is

```
/__w/json/json/single_include/nlohmann/json.hpp:19046:65: error: template argument for template type parameter must be a type
                       detail::negation<std::is_same<ValueType, nullptr_t>>,
                                                                ^~~~~~~~~
```

Any idea?

--- a/cmake/ci.cmake
+++ b/cmake/ci.cmake
@@ -30,7 +30,7 @@ execute_process(COMMAND ${CPPCHECK_TOOL} --version OUTPUT_VARIABLE CPPCHECK_TOOL
 string(REGEX MATCH "[0-9]+(\\.[0-9]+)+" CPPCHECK_TOOL_VERSION "${CPPCHECK_TOOL_VERSION}")
 message(STATUS "🔖 Cppcheck ${CPPCHECK_TOOL_VERSION} (${CPPCHECK_TOOL})")
 
-find_program(GCC_TOOL NAMES g++-HEAD g++-11 g++-latest)
+find_program(GCC_TOOL NAMES g++-latest g++-HEAD g++-11)
 execute_process(COMMAND ${GCC_TOOL} --version OUTPUT_VARIABLE GCC_TOOL_VERSION ERROR_VARIABLE GCC_TOOL_VERSION)
 string(REGEX MATCH "[0-9]+(\\.[0-9]+)+" GCC_TOOL_VERSION "${GCC_TOOL_VERSION}")
 message(STATUS "🔖 GCC ${GCC_TOOL_VERSION} (${GCC_TOOL})")

--- a/cmake/ci.cmake
+++ b/cmake/ci.cmake
@@ -111,6 +111,7 @@ set(CLANG_CXXFLAGS "-std=c++11                        \
     -Wno-reserved-identifier                             \
 ")
 
+# Warning flags determined for GCC 12.0 (experimental) with https://github.com/nlohmann/gcc_flags:
 # Ignored GCC warnings:
 # -Wno-abi-tag                    We do not care about ABI tags.
 # -Wno-aggregate-return           The library uses aggregate returns.
@@ -150,16 +151,22 @@ set(GCC_CXXFLAGS "-std=c++11                          \
     -Wanalyzer-shift-count-negative                   \
     -Wanalyzer-shift-count-overflow                   \
     -Wanalyzer-stale-setjmp-buffer                    \
+    -Wanalyzer-tainted-allocation-size                \
     -Wanalyzer-tainted-array-index                    \
+    -Wanalyzer-tainted-divisor                        \
+    -Wanalyzer-tainted-offset                         \
+    -Wanalyzer-tainted-size                           \
     -Wanalyzer-too-complex                            \
     -Wanalyzer-unsafe-call-within-signal-handler      \
     -Wanalyzer-use-after-free                         \
     -Wanalyzer-use-of-pointer-in-stale-stack-frame    \
+    -Wanalyzer-use-of-uninitialized-value             \
     -Wanalyzer-write-to-const                         \
     -Wanalyzer-write-to-string-literal                \
     -Warith-conversion                                \
     -Warray-bounds                                    \
     -Warray-bounds=2                                  \
+    -Warray-compare                                   \
     -Warray-parameter=2                               \
     -Wattribute-alias=2                               \
     -Wattribute-warning                               \
@@ -170,10 +177,15 @@ set(GCC_CXXFLAGS "-std=c++11                          \
     -Wbuiltin-macro-redefined                         \
     -Wc++0x-compat                                    \
     -Wc++11-compat                                    \
+    -Wc++11-extensions                                \
     -Wc++14-compat                                    \
+    -Wc++14-extensions                                \
     -Wc++17-compat                                    \
+    -Wc++17-extensions                                \
     -Wc++1z-compat                                    \
     -Wc++20-compat                                    \
+    -Wc++20-extensions                                \
+    -Wc++23-extensions                                \
     -Wc++2a-compat                                    \
     -Wcannot-profile                                  \
     -Wcast-align                                      \
@@ -191,6 +203,7 @@ set(GCC_CXXFLAGS "-std=c++11                          \
     -Wconditionally-supported                         \
     -Wconversion                                      \
     -Wconversion-null                                 \
+    -Wcoverage-invalid-line-number                    \
     -Wcoverage-mismatch                               \
     -Wcpp                                             \
     -Wctad-maybe-unsupported                          \
@@ -215,21 +228,16 @@ set(GCC_CXXFLAGS "-std=c++11                          \
     -Wendif-labels                                    \
     -Wenum-compare                                    \
     -Wenum-conversion                                 \
+    -Wexceptions                                      \
     -Wexpansion-to-defined                            \
     -Wextra                                           \
     -Wextra-semi                                      \
     -Wfloat-conversion                                \
     -Wfloat-equal                                     \
-    -Wformat-contains-nul                             \
     -Wformat-diag                                     \
-    -Wformat-extra-args                               \
-    -Wformat-nonliteral                               \
     -Wformat-overflow=2                               \
-    -Wformat-security                                 \
     -Wformat-signedness                               \
     -Wformat-truncation=2                             \
-    -Wformat-y2k                                      \
-    -Wformat-zero-length                              \
     -Wformat=2                                        \
     -Wframe-address                                   \
     -Wfree-nonheap-object                             \
@@ -239,12 +247,15 @@ set(GCC_CXXFLAGS "-std=c++11                          \
     -Wignored-qualifiers                              \
     -Wimplicit-fallthrough=5                          \
     -Winaccessible-base                               \
+    -Winfinite-recursion                              \
     -Winherited-variadic-ctor                         \
     -Winit-list-lifetime                              \
     -Winit-self                                       \
     -Winline                                          \
     -Wint-in-bool-context                             \
     -Wint-to-pointer-cast                             \
+    -Winterference-size                               \
+    -Winvalid-imported-macros                         \
     -Winvalid-memory-model                            \
     -Winvalid-offsetof                                \
     -Winvalid-pch                                     \
@@ -267,6 +278,7 @@ set(GCC_CXXFLAGS "-std=c++11                          \
     -Wmissing-field-initializers                      \
     -Wmissing-include-dirs                            \
     -Wmissing-profile                                 \
+    -Wmissing-requires                                \
     -Wmultichar                                       \
     -Wmultiple-inheritance                            \
     -Wmultistatement-macros                           \
@@ -282,6 +294,7 @@ set(GCC_CXXFLAGS "-std=c++11                          \
     -Wnull-dereference                                \
     -Wodr                                             \
     -Wold-style-cast                                  \
+    -Wopenacc-parallelism                             \
     -Wopenmp-simd                                     \
     -Woverflow                                        \
     -Woverlength-strings                              \

--- a/include/nlohmann/json.hpp
+++ b/include/nlohmann/json.hpp
@@ -1872,6 +1872,7 @@ class basic_json // NOLINT(cppcoreguidelines-special-member-functions,hicpp-spec
     template < typename ValueType, typename std::enable_if <
                    detail::conjunction <
                        detail::negation<std::is_pointer<ValueType>>,
+                       detail::negation<std::is_same<ValueType, std::nullptr_t>>,
                        detail::negation<std::is_same<ValueType, detail::json_ref<basic_json>>>,
                                         detail::negation<std::is_same<ValueType, typename string_t::value_type>>,
                                         detail::negation<detail::is_basic_json<ValueType>>,

--- a/include/nlohmann/ordered_map.hpp
+++ b/include/nlohmann/ordered_map.hpp
@@ -30,7 +30,8 @@ template <class Key, class T, class IgnoredLess = std::less<Key>,
 
     // Explicit constructors instead of `using Container::Container`
     // otherwise older compilers choke on it (GCC <= 5.5, xcode <= 9.4)
-    ordered_map(const Allocator& alloc = Allocator()) : Container{alloc} {}
+    ordered_map() noexcept(noexcept(Container())) : Container{} {}
+    explicit ordered_map(const Allocator& alloc) noexcept(noexcept(Container(alloc))) : Container{alloc} {}
     template <class It>
     ordered_map(It first, It last, const Allocator& alloc = Allocator())
         : Container{first, last, alloc} {}

--- a/single_include/nlohmann/json.hpp
+++ b/single_include/nlohmann/json.hpp
@@ -19052,6 +19052,7 @@ class basic_json // NOLINT(cppcoreguidelines-special-member-functions,hicpp-spec
     template < typename ValueType, typename std::enable_if <
                    detail::conjunction <
                        detail::negation<std::is_pointer<ValueType>>,
+                       detail::negation<std::is_same<ValueType, std::nullptr_t>>,
                        detail::negation<std::is_same<ValueType, detail::json_ref<basic_json>>>,
                                         detail::negation<std::is_same<ValueType, typename string_t::value_type>>,
                                         detail::negation<detail::is_basic_json<ValueType>>,

--- a/single_include/nlohmann/json.hpp
+++ b/single_include/nlohmann/json.hpp
@@ -17066,7 +17066,8 @@ template <class Key, class T, class IgnoredLess = std::less<Key>,
 
     // Explicit constructors instead of `using Container::Container`
     // otherwise older compilers choke on it (GCC <= 5.5, xcode <= 9.4)
-    ordered_map(const Allocator& alloc = Allocator()) : Container{alloc} {}
+    ordered_map() noexcept(noexcept(Container())) : Container{} {}
+    explicit ordered_map(const Allocator& alloc) noexcept(noexcept(Container(alloc))) : Container{alloc} {}
     template <class It>
     ordered_map(It first, It last, const Allocator& alloc = Allocator())
         : Container{first, last, alloc} {}

--- a/test/src/unit-allocator.cpp
+++ b/test/src/unit-allocator.cpp
@@ -39,6 +39,8 @@ namespace
 template<class T>
 struct bad_allocator : std::allocator<T>
 {
+    using std::allocator<T>::allocator;
+
     template<class... Args>
     void construct(T* /*unused*/, Args&& ... /*unused*/)
     {

--- a/test/src/unit-alt-string.cpp
+++ b/test/src/unit-alt-string.cpp
@@ -104,12 +104,12 @@ class alt_string
     }
 
     template <typename op_type>
-    bool operator<(const op_type& op) const
+    bool operator<(const op_type& op) const noexcept
     {
         return str_impl < op;
     }
 
-    bool operator<(const alt_string& op) const
+    bool operator<(const alt_string& op) const noexcept
     {
         return str_impl < op.str_impl;
     }

--- a/test/src/unit-comparison.cpp
+++ b/test/src/unit-comparison.cpp
@@ -32,6 +32,10 @@ SOFTWARE.
 #include <nlohmann/json.hpp>
 using nlohmann::json;
 
+#if (defined(__cplusplus) && __cplusplus >= 202002L) || (defined(_MSVC_LANG) && _MSVC_LANG >= 202002L)
+    #define JSON_HAS_CPP_20
+#endif
+
 namespace
 {
 // helper function to check std::less<json::value_t>
@@ -205,6 +209,14 @@ TEST_CASE("lexicographical comparison operators")
             {
                 for (size_t j = 0; j < j_values.size(); ++j)
                 {
+                    // Skip comparing indicies 12 and 13, and 13 and 12 in C++20 pending fix
+                    // See issue #3207
+#ifdef JSON_HAS_CPP_20
+                    if ((i == 12 && j == 13) || (i == 13 && j == 12))
+                    {
+                        continue;
+                    }
+#endif
                     CAPTURE(i)
                     CAPTURE(j)
                     CAPTURE(j_values[i])

--- a/test/src/unit-regression2.cpp
+++ b/test/src/unit-regression2.cpp
@@ -341,7 +341,7 @@ TEST_CASE("regression tests 2")
                ]
              })";
 
-        json::parser_callback_t cb = [&](int /*level*/, json::parse_event_t event, json & parsed)
+        json::parser_callback_t cb = [&](int /*level*/, json::parse_event_t event, json & parsed) noexcept
         {
             // skip uninteresting events
             if (event == json::parse_event_t::value && !parsed.is_primitive())

--- a/test/src/unit-regression2.cpp
+++ b/test/src/unit-regression2.cpp
@@ -271,7 +271,10 @@ std::string* sax_no_exception::error_string = nullptr;
 
 template<class T>
 class my_allocator : public std::allocator<T>
-{};
+{
+  public:
+    using std::allocator<T>::allocator;
+};
 
 /////////////////////////////////////////////////////////////////////
 // for #3077

